### PR TITLE
[Feature] Skip existing for advantage modules

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by albertbou92** [comment]
What about the other way around? by default not skipping.

In that case we should make sure in the advantage module the value is computed without gradients.

**Comment by matteobettini** [comment]
Could you explain a bit more why your example constitutes a problem?
Wouldn't the value computed during rollout be the same if we just compute it later in the loss?

Also I wonder if this is a small step in the direction of #1000.
Because here we are allowing to precompute values  (maybe on bigger batches) outside losse forward.
And in #1000 we want to allow also precomputing targets (maybe on bigger batches) outside loss forward.

**Comment by vmoens** [comment]
> Could you explain a bit more why your example constitutes a problem? Wouldn't the value computed during rollout be the same if we just compute it later in the loss?
> 
> Also I wonder if this is a small step in the direction of #1000. Because here we are allowing to precompute values (maybe on bigger batches) outside losse forward. And in #1000 we want to allow also precomputing targets (maybe on bigger batches) outside loss forward.
> 
> PS in #1000 advantages would not even have access to neural networks. Only losses will. and thus it will make all this even simpler

you can perfectly not pass a network and it will work fine.
I'd like to keep the networks bc it's much easier, otherwise we are decomposing things and we'll be asking users to do
```
value_net(td)
value_net(td.get("next"))
gae(td)
```
instead of 
```
gae(td)
```

**Comment by matteobettini** [comment]
Yep i didn't see that we could pass None, that works perfectly.

Thinking about this more, if losses start skipping existing keys by default, this may have some bad consequences if collected data (from collectors or buffers) has some of the loss keys of interest (cause they might be old).

But thinking of #1000 we definitely need a way to precomupte keys for losses (values and targets) and make losses use those precomputed keys.

I think it would be fine to use that skip_existing by default also in the losses and if users feed the losses data with old values it is their problem.
```
batch = rb.sample(60_000)
loss_module.compute_value_target(batch)
for _ in range(n):
    minibatch = subsample(batch, 1000)
    loss_vals = loss_module(minibatch) # by default uses the precomputed targets
```

Alternatively, what @albertbou92 says also makes sense, the losses could recompute everything by default and we would need to
```
batch = rb.sample(60_000)
loss_module.compute_value_target(batch)
for _ in range(n):
    minibatch = subsample(batch, 1000)
    loss_vals = loss_module(minibatch, skip_exisitng=True) # we need to enforce it to not recompute the targets
```

I am leaning more towards the first option tho

**Comment by vmoens** [comment]
> Thinking about this more, if losses start skipping existing keys by default, this may have some bad consequences if collected data (from collectors or buffers) has some of the loss keys of interest (cause they might be old).

yes this is why I set it to False, and why I was asking you guys what you were thinking. I guess we'll leave it as it is.

> But thinking of #1000 we definitely need a way to precomupte keys for losses (values and targets) and make losses use those precomputed keys.

if you don't pass the value net to the module an error will be raised bc the entry is missing.

> I think it would be fine to use that skip_existing by default also in the losses and if users feed the losses data with old values it is their problem.

yep that's the idea.
There are 2 things you could do
```
with set_skip_existing(True):
    loss_fn(td)
```
or 
```
loss_fn.make_value_estimator(..., skip_existing=True)
```

**Comment by matteobettini** [comment]
I am more for the first of the 2 since you are not bound to smth at init time.

Regarding the default of skip_existing i am leaning more towards True but I am fine with anything

**Comment by vmoens** [comment]
> I am more for the first of the 2 since you are not bound to smth at init time.

then use that one 😉


> Regarding the default of skip_existing i am leaning more towards True but I am fine with anything

Let's put it that way:
worst that can happen when True: wrong value -> no training -> torchrl sucks
worst that can happen when False: slow training -> good training, but slow -> issue

**Comment by albertbou92** [comment]
> > I am more for the first of the 2 since you are not bound to smth at init time.
> 
> then use that one 😉
> 
> > Regarding the default of skip_existing i am leaning more towards True but I am fine with anything
> 
> Let's put it that way: worst that can happen when True: wrong value -> no training -> torchrl sucks worst that can happen when False: slow training -> good training, but slow -> issue

Yes  I like the idea of having something that works out of the box, even if there is later on margin for improvement in terms of performance and requires being a bit of an advanced user
